src/boot/efi/cpio.c

```diff
@@ -0,0 +1,465 @@
+        _cleanup_freepool_ CHAR16 *loaded_image_path = NULL, *j = NULL;
+        _cleanup_(strv_freep) CHAR16 **items = NULL;
+        _cleanup_freepool_ VOID *buffer = NULL;
+        UINT32 inode = 1; /* inode counter, so that each item gets a new inode */
+        EFI_STATUS err;
+
+        assert(loaded_image);
+        assert(ret_buffer);
+        assert(ret_buffer_size);
+
+        root = LibOpenRoot(loaded_image->DeviceHandle);
+        if (!root)
+                return log_error_status_stall(EFI_LOAD_ERROR, L"Unable to open root directory.");
+
+        loaded_image_path = DevicePathToStr(loaded_image->FilePath);
+        if (!loaded_image_path)
+                return log_oom();
+
+        l = StrLen(loaded_image_path);
+        a = StrLen(EXTRA_DIR_SUFFIX);
+        if (l >= UINTN_MAX - a)
+                return log_error_status_stall(EFI_OUT_OF_RESOURCES, L"Path length overflow.");
+
+        j = AllocatePool(sizeof(CHAR16) * (l + a + 1));
+        if (!j)
+                return log_oom();
+
+        StpCpy(StpCpy(j, loaded_image_path), EXTRA_DIR_SUFFIX);
+
+        err = uefi_call_wrapper(root->Open, 5, root, &extra_dir, j, EFI_FILE_MODE_READ, 0ULL);
+        if (err == EFI_NOT_FOUND) {
+                /* No extra subdir, that's totally OK */
+                *ret_buffer = NULL;
+                *ret_buffer_size = 0;
+                return EFI_SUCCESS;
+        }
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed to open extra directory of loaded image: %r", err);
+
+        /* We opened the extra directory now, let's verify it actually *is* a directory */
+
+        err = get_file_info_harder(extra_dir, &file_info);
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed to get information about extra directory of loaded image: %r", err);
+        if (!(file_info->Attribute & EFI_FILE_DIRECTORY))
+                return log_error_status_stall(EFI_NOT_FOUND, L"Extra initrd directory is not actually a directory, refusing.");
+
+        for (;;) {
+                _cleanup_freepool_ CHAR16 *d = NULL;
+
+                err = readdir_harder(extra_dir, &dirent, &dirent_size);
+                if (EFI_ERROR(err))
+                        return log_error_status_stall(err, L"Failed to read extra directory of loaded image: %r", err);
+                if (!dirent) /* End of directory */
+                        break;
+
+                if (dirent->FileName[0] == '.')
+                        continue;
+                if (dirent->Attribute & EFI_FILE_DIRECTORY)
+                        continue;
+                if (match_suffix && !endswith_no_case(dirent->FileName, match_suffix))
+                        continue;
+                if (!is_ascii(dirent->FileName))
+                        continue;
+                if (StrLen(dirent->FileName) > 255) /* Max filename size on Linux */
+                        continue;
```

Comment:
```
                if (dirent->FileName[0] == '.' ||
                    dirent->Attribute & EFI_FILE_DIRECTORY ||
                    match_suffix && !endswith_no_case(dirent->FileName, match_suffix) ||
                    !is_ascii(dirent->FileName) ||
                    StrLen(dirent->FileName) > 255) /* Max filename size on Linux */
                        continue;
```
Feels nicer to me, not that it matters.

Reply:
not a fan of this dense style, and it requires a bunch of more ()...